source/agora/cli/multi/main.d

```diff
@@ -0,0 +1,329 @@
+/*******************************************************************************
+
+    Entry point for the Agora multi
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.cli.multi.main;
+
+import agora.api.FullNode;
+
+import agora.common.Config;
+import agora.node.Node;
+import agora.utils.Log;
+
+import vibe.core.core;
+import vibe.web.rest;
+
+import std.exception;
+import std.file;
+import std.format;
+import std.path;
+import std.range;
+import std.stdio;
+import std.string;
+
+
+/// Default port for nodes when forwarding ports
+private immutable ushort NODE_BASE_PORT = 4000;
+
+/// Admin port for nodes when forwarding ports
+private immutable ushort ADMIN_BASE_PORT = 5000;
+
+/*******************************************************************************
+
+    Application entry point
+
+*******************************************************************************/
+
+private int main (string[] args)
+{
+    // Validate parameters
+    if ((args.length < 3) || ((args[1] != "start") && (args[1] != "test")))
+    {
+        printDefaultHelp();
+        return 1;
+    }
+
+    string[] files;
+    // If the root path is not set,
+    if (extension(args[2]).toLower == ".yaml")
+        files = args[2..$];
+    else
+    // If the root path is not set,
+        foreach (idx; 3..args.length)
+            files ~= buildNormalizedPath(args[2], args[idx]);
+
+    // Reads all configurations from the file.
+    Config[] configs;
+    foreach (idx, file; files)
+    {
+        try
+        {
+            // Read
+            auto config = readFromFile(file);
+
+            // Convert data to suit the environment.
+
+            // Change the IP of the server to remove the pop-up window when using the firewall.
+            config.node.address = "127.0.0.1";
+
+            // The configuration file is supposed to run inside the dock.
+            // Ports on nodes should not overlap each other.
+            // Ports 4000, 4001,4002 are used by `tests/system/source/main.d.`
+            // Change the port to match these rules.
+            config.node.port = cast(ushort)(NODE_BASE_PORT + idx);
```

Comment:
Oh I see why you're doing this now!
Let's just change the test to take the addresses as arguments.

Reply:
Then I will modify the `tests/system/source/main.d` to get the addresses as an argument.